### PR TITLE
[AOT] Add RequiresDynamicCode annotation to Microsoft.AspNetCore.Routing

### PR DESCRIPTION
**Pull request by Nick-Stanton**
# [AOT] Add RequiresDynamicCode annotation to Microsoft.AspNetCore.Routing

Adds `RequiresDynamicCode` annotations and appropriate warning suppressions to prepare Web UI's projects for AOT analysis.

## Description

The addition of the AOT analysis into our projects brought in ~40 warnings for us, half of which were related to reflection. The other half consist of JSON warnings (#45527) and Dependency Injection (dotnet/runtime#79286) and will be handled in the linked issues. This PR addresses the reflection warnings with the exception of `BindConverter.cs` which requires further conversation (see #45578). Please note that reflection with `MakeGenericType` and `MakeGenericMethod` are allowed for reference types and not value types in AOT, so suppressions should _only_ happen in cases where we know only a reference type is being passed.

Part of effort tracked in #45473

### CODE REVIEW

File: src/Components/Components/src/Reflection/PropertySetter.cs [hunk 1]
```diff
@@ -17,6 +17,10 @@ internal sealed class PropertySetter
         "ReflectionAnalysis",
         "IL2060:MakeGenericMethod",
         Justification = "The referenced methods don't have any DynamicallyAccessedMembers annotations. See https://github.com/mono/linker/issues/1727")]
+    [SuppressMessage(
+        "AOT",
+        "IL3050:Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.",
+        Justification = "The referenced methods are AOT safe with reference types.")]
```
Comment (JamesNK) [inline src/Components/Components/src/Reflection/PropertySetter.cs:23]: Is this only used with reference types?

Add a `Debug.Assert(targetType.IsClass)` and `Debug.Assert(property.PropertyType.IsClass)` to verify.

File: src/Http/Routing/src/Matching/JumpTableBuilder.cs [hunk 2]
```diff
@@ -87,7 +87,9 @@ public static JumpTable Build(int defaultDestination, int exitDestination, (stri
         // Use the ILEmitTrieJumpTable if the IL is going to be compiled (not interpreted)
         if (RuntimeFeature.IsDynamicCodeCompiled)
         {
+#pragma warning disable IL3050 // See https://github.com/dotnet/linker/issues/2715.
             return new ILEmitTrieJumpTable(defaultDestination, exitDestination, pathEntries, vectorize: null, fallback);
```
Comment (JamesNK) [inline src/Http/Routing/src/Matching/JumpTableBuilder.cs:91]: Hmm. AOT analysis recognizes pragma disable, but I don't believe it flows through to the AOT compiler.

Instead, create a local method, move `ILEmitTrieJumpTable` usage to it, and place the attribute on the local method.

File: src/Components/Components/src/Reflection/PropertySetter.cs (hunk 1 above)
Comment (JamesNK) [reply src/Components/Components/src/Reflection/PropertySetter.cs:23]: You've used SuppressMessage instead of UnconditionalSuppressMessage throughout this PR. Is there a reason why?

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetObjectReferenceJsonConverterFactory.cs [hunk 3]
```diff
@@ -22,6 +22,7 @@ public override bool CanConvert(Type typeToConvert)
     }
 
     [UnconditionalSuppressMessage("Trimming", "IL2055", Justification = "We expect that types used with DotNetObjectReference are retained.")]
+    [SuppressMessage("AOT", "IL3050:Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.", Justification = "MakeGenericType is AOT safe for reference types.")]
```
Comment (JamesNK) [inline src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetObjectReferenceJsonConverterFactory.cs:25]: Is this always used with reference types?

Add a `Debug.Assert(instanceType.IsClass)` to verify.

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/TaskGenericsUtil.cs [hunk 4]
```diff
@@ -23,6 +24,9 @@ public static void SetTaskCompletionSourceException(object taskCompletionSource,
     public static Type GetTaskCompletionSourceResultType(object taskCompletionSource)
         => CreateResultSetter(taskCompletionSource).ResultType;
 
+    [SuppressMessage("AOT",
+        "IL3050:Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.",
+        Justification = "MakeGenericType is AOT safe for reference types.")]
```
Comment (JamesNK) [inline src/JSInterop/Microsoft.JSInterop/src/Infrastructure/TaskGenericsUtil.cs:29]: Add Debug.Assert to verify result task is reference type

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/TaskGenericsUtil.cs [hunk 5]
```diff
@@ -101,6 +105,9 @@ public void SetException(object tcs, Exception exception)
         }
     }
 
+    [SuppressMessage("AOT",
+        "IL3050:Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.",
+        Justification = "MakeGenericType is AOT safe for reference types.")]
```
Comment (JamesNK) [inline src/JSInterop/Microsoft.JSInterop/src/Infrastructure/TaskGenericsUtil.cs:110]: Same as above

File: src/Shared/PropertyHelper/PropertyHelper.cs [hunk 6]
```diff
@@ -151,6 +151,7 @@ public void SetValue(object instance, object? value)
     /// A cached array of all public properties of the specified type.
     /// </returns>
     [RequiresUnreferencedCode("This API is not trim safe.")]
+    [RequiresDynamicCode("This API is not trim safe.")]
```
Comment (JamesNK) [inline src/Shared/PropertyHelper/PropertyHelper.cs:154]: Could you move `"This API is not trim safe."` to a const on this type and reuse in all the attributes. Some are inconsistent.

File: src/Components/Components/src/Reflection/PropertySetter.cs (hunk 1 above)
Comment (Nick-Stanton) [reply src/Components/Components/src/Reflection/PropertySetter.cs:23]: > You've used SuppressMessage instead of UnconditionalSuppressMessage throughout this PR. Is there a reason why?

Nope, just what VS suggested. I'll fix those and add the assertions.

Comment (eerhardt) [inline src/Components/Components/src/Reflection/PropertySetter.cs:23]: I don't think this suppression is valid. We won't ever call this class with a property of type `int`, `double`, or `Guid`, or any other Value Type?

Instead, if we want to make this AOT-compatible, we will need to do something like this:

https://github.com/dotnet/runtime/blob/66b11d7f2c2b25fc3e6685b02f016ef2f9f8c64f/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L1378-L1391

Basically - check if `!RuntimeFeature.IsDynamicCodeSupported`. When it isn't, and one of the types is a ValueType, then fallback to "slower" reflection based property setting that boxes.

File: src/Http/Routing/src/Matching/JumpTableBuilder.cs (hunk 2 above)
Comment (eerhardt) [reply src/Http/Routing/src/Matching/JumpTableBuilder.cs:91]: This is actually OK, because it is working around a bug in the AOT analyzer.

When the AOT compiler runs, the code inside the `if` will be trimmed away (because `IsDynamicCodeCompiled` will never be true). So the AOT compiler won't warn, since the code is trimmed.

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs [hunk 7]
```diff
@@ -375,6 +375,9 @@ private static (MethodInfo, Type[]) GetCachedMethodInfo(AssemblyKey assemblyKey,
         "ReflectionAnalysis",
         "IL2060:MakeGenericMethod",
         Justification = "https://github.com/mono/linker/issues/1727")]
```
Comment (eerhardt) [inline src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:377]: This issue is fixed. Can we remove this existing suppression while you are in here?

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs [hunk 8]
```diff
@@ -375,6 +375,9 @@ private static (MethodInfo, Type[]) GetCachedMethodInfo(AssemblyKey assemblyKey,
...
+    [SuppressMessage("AOT",
+        "IL3050:Calling members annotated with 'RequiresDynamicCodeAttribute' may break functionality when AOT compiling.",
+        Justification = "Methods being referenced do not have DynamicallyAccessedMembers.")]
```
Comment (eerhardt) [inline src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:380]: 1. This isn't a valid justification for suppressing this warning.
2. I don't believe this warning can be safely suppressed. From looking at the only callsite, this is used to convert a `ValueTask<T>` to a `Task<T>`. If the `T` is a ValueType, this code isn't guaranteed to work in Native AOT, since `CreateValueTaskConverter<T>` might not be generated ahead of time for that specific ValueType.

Comment (eerhardt) [reply src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:380]: Further, how important is it for Microsoft.JSInterop to be AOT-compatible in .NET 8? Isn't it only used for WASM, which doesn't use Native AOT?

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/TaskGenericsUtil.cs (hunk 4 above)
Comment (eerhardt) [inline src/JSInterop/Microsoft.JSInterop/src/Infrastructure/TaskGenericsUtil.cs:29]: I don't think either warnings being suppressed in this file are valid. Why are the `Tasks` being used here guaranteed to be of reference types?

File: src/Shared/PropertyHelper/PropertyHelper.cs (hunk 6 above)
Comment (eerhardt) [inline src/Shared/PropertyHelper/PropertyHelper.cs:154]:
```suggestion
    [RequiresDynamicCode("This API requires dynamic code because it makes generic types which may be filled with ValueTypes.")]
```

`RequiresDynamicCode` != `not trim safe`. Instead, it means that this API may need to generate and JIT compile code, which is not allowed in Native AOT.

Comment (JamesNK) [inline src/Shared/PropertyHelper/PropertyHelper.cs:154]: I looked at PropertyHelper.cs in my PR here - https://github.com/dotnet/aspnetcore/pull/45604/files#diff-8d64fa5ac96afa5f91aa693105c4f410335bfaa79ef790546ae1773e3f50f594

Instead of adding `RequiresDynamicCodeAttribute`, I removed dynamic code to avoid the attribute spreading everywhere.

Comment (Nick-Stanton) [reply src/Shared/PropertyHelper/PropertyHelper.cs:154]: Thanks James!

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs (hunk 8 above)
Comment (Nick-Stanton) [reply src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs]: It is used across WASM, Server and WebView. You've raised important points about these suppressions, and I will either pursue a fallback like you describe [here](https://github.com/dotnet/aspnetcore/pull/45580#discussion_r1048741270) or document my findings in an issue report for us to milestone and discuss like I did with BindConverter (#45578).

File: src/Shared/PropertyHelper/PropertyHelper.cs (hunk 6 above)
Comment (Nick-Stanton) [reply src/Shared/PropertyHelper/PropertyHelper.cs]: Resolved by #45604

**Comment by Nick-Stanton** [comment]
Upon further investigation, more work needs to be done than just suppressing these warnings. The scope of this PR is pivoting to labeling what we already know will be unsafe.

#45578 and #45795 track the remaining work for Web UI to complete as part of the AOT analysis warning effort.

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs (hunk 8 above)
Comment (Nick-Stanton) [reply src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs]: Documented in #45795

**Review by MackinnonBuck — APPROVED** [review]
